pkg/client/events_query/client.go

```diff
@@ -0,0 +1,289 @@
+
+// TODO_CONSIDERATION: the cosmos-sdk CLI code seems to use a cometbft RPC client
+// which includes a `#EventsBytes()` method for a similar purpose. Perhaps we could
+// replace this custom websocket client with that.
+// (see: https://github.com/cometbft/cometbft/blob/main/rpc/client/http/http.go#L110)
+// (see: https://github.com/cosmos/cosmos-sdk/blob/main/client/rpc/tx.go#L114)
+
+// eventsQueryClient implements the EventsQueryClient interface.
+type eventsQueryClient struct {
+	// cometWebsocketURL is the websocket URL for the cometbft node. It is assigned
+	// in NewEventsQueryClient.
+	cometWebsocketURL string
+	// nextRequestId is a *unique* ID intended to be monotonically incremented
+	// and used to uniquely identify distinct RPC requests.
+	// TODO_CONSIDERATION: Consider changing `nextRequestId` to a random entropy field
+	nextRequestId uint64
+
+	// dialer is resopnsible for createing the connection instance which
+	// facilitates communication with the cometbft node via message passing.
+	dialer client.Dialer
+	// eventsBytesAndConnsMu protects the eventsBytesAndConns map.
+	eventsBytesAndConnsMu sync.RWMutex
+	// eventsBytesAndConns maps event subscription queries to their respective
+	// eventsBytes observable, connection, and closed status.
+	eventsBytesAndConns map[string]*eventsBytesAndConn
+}
+
+// eventsBytesAndConn is a struct which holds an eventsBytes observable & the
+// corresponding connection which produces its inputs.
+type eventsBytesAndConn struct {
+	// eventsBytes is an observable which is notified about chain event messages
+	// matching the given query. It receives an either.Either[[]byte] which is
+	// either an error or the event message bytes.
+	eventsBytes observable.Observable[either.Either[[]byte]]
+	conn        client.Connection
+	closed      bool
+}
+
+func NewEventsQueryClient(cometWebsocketURL string, opts ...client.EventsQueryClientOption) client.EventsQueryClient {
+	evtClient := &eventsQueryClient{
+		cometWebsocketURL:   cometWebsocketURL,
+		eventsBytesAndConns: make(map[string]*eventsBytesAndConn),
+	}
+
+	for _, opt := range opts {
+		opt(evtClient)
+	}
+
+	if evtClient.dialer == nil {
+		// default to using the websocket dialer
+		evtClient.dialer = NewWebsocketDialer()
+	}
+
+	return evtClient
+}
+
+// EventsBytes returns an eventsBytes observable which is notified about chain
+// event messages matching the given query. It receives an either.Either[[]byte]
+// which is either an error or the event message bytes.
+// (see: https://pkg.go.dev/github.com/cometbft/cometbft/types#pkg-constants)
+// (see: https://docs.cosmos.network/v0.47/core/events#subscribing-to-events)
+func (eqc *eventsQueryClient) EventsBytes(
+	ctx context.Context,
+	query string,
+) (client.EventsBytesObservable, error) {
+	// Check if an event subscription already exists for the given query.
+	if eventsBzConn := eqc.getEventsBytesAndConn(query); eventsBzConn != nil {
```

Comment:
I think there should be a lock should wrapping the code from `getEventsBytesAndConn` til `insertEventBytesAndConn` otherwise concurrent calls to `EventsBytes` may both create and insert for the same `query`

Reply:
Both `getEventsBytesAndConn()` and `insertEventsBytesAndConn()` acquire a lock on the mutex before reading  or writing `eventsBytesAndCons`.